src/zarr/group.py

```diff
@@ -49,13 +50,15 @@ class GroupMetadata(Metadata):
     node_type: Literal["group"] = field(default="group", init=False)
 
     # todo: rename this, since it doesn't return bytes
-    def to_bytes(self) -> dict[str, bytes]:
+    def to_bytes(self) -> dict[str, bytes | None]:
         if self.zarr_format == 3:
             return {ZARR_JSON: json.dumps(self.to_dict()).encode()}
         else:
             return {
                 ZGROUP_JSON: json.dumps({"zarr_format": 2}).encode(),
-                ZATTRS_JSON: json.dumps(self.attributes).encode(),
+                ZATTRS_JSON: json.dumps(self.attributes).encode()
+                if len(self.attributes) > 0
+                else None,
```

Comment:
why make this function nullable?

Reply:
To allow that `.zattrs` can be deleted if the attributes are empty.

Reply:
can't we do that by checking if the attributes are the empty dict `{}`?

Reply:
Sure. But that would break the abstraction a bit.

Reply:
which abstraction does it break?

Reply:
I think a special-casing serializing `{}` to JSON as a deletion is a fine price to pay for a simpler (non-nullable) type signature

Reply:
> which abstraction does it break?

The `GroupMetadata` class provides the bytes to store away without needing to know about the bytes contents or semantics.

Reply:
what if we just omit ZATTRS_JSON from this dict if the attributes are empty?

Reply:
Then the `Group` needs to know about the file names of the zarr formats (which it already does for opening).

Reply:
alternatively, we could just write the `{}` to storage and not worry about deleting anything. Readers are going to try and fetch that file anyway, so I'm not sure there's a performance difference between an absent JSON file and one that contains `{}`

Reply:
Fine with me.